File: ios/Wavepoint/Cleanup/CleanupDeckBuilder.swift
import Foundation

struct CleanupDeckBuilder: Sendable {
  let maximumTrackCount: Int
  let referenceDate: Date

  init(
    maximumTrackCount: Int = 50,
    referenceDate: Date = .now
  ) {
    self.maximumTrackCount = maximumTrackCount
    self.referenceDate = referenceDate
  }

  func build(
    from tracks: [LibraryTrack],
    maximumTrackCount: Int? = nil,
    seed: UInt64 = UInt64.random(in: UInt64.min...UInt64.max)
  ) -> [LibraryTrack] {
    var generator = SplitMix64(seed: seed)
    let trackLimit = max(0, maximumTrackCount ?? self.maximumTrackCount)

    return
      tracks
      .map { track in
        let randomValue = max(
          Double.random(in: 0..<1, using: &generator),
          Double.leastNonzeroMagnitude
        )
        let weight = selectionWeight(for: track)
        return (track: track, priority: log(randomValue) / weight)
      }
      .sorted { lhs, rhs in
        if lhs.priority != rhs.priority {
          return lhs.priority > rhs.priority
        }
        return lhs.track.id < rhs.track.id
      }
      .prefix(trackLimit)
      .map(\.track)
  }

  func selectionWeight(for track: LibraryTrack) -> Double {
    let ageInYears = track.addedAt.map {
      max(0, referenceDate.timeIntervalSince($0) / (365.25 * 24 * 60 * 60))
    } ?? 0
    let ageWeight = 1 + min(ageInYears, 12)
    return ageWeight
  }
}

private struct SplitMix64: RandomNumberGenerator {
  private var state: UInt64

  init(seed: UInt64) {
    state = seed
  }

  mutating func next() -> UInt64 {
    state &+= 0x9E37_79B9_7F4A_7C15
    var value = state
    value = (value ^ (value >> 30)) &* 0xBF58_476D_1CE4_E5B9
    value = (value ^ (value >> 27)) &* 0x94D0_49BB_1331_11EB
    return value ^ (value >> 31)
  }
}

File: android/app/src/main/java/ai/mapier/swipe/cleanup/CleanupDeckBuilder.kt
package ai.mapier.swipe.cleanup

import java.time.Duration
import java.time.Instant
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.min

class CleanupDeckBuilder(
  private val maximumTrackCount: Int = 50,
  private val referenceDate: Instant = Instant.now(),
) {
  fun build(
    tracks: List<LibraryTrack>,
    recentTrackIds: Set<String>,
    seed: Long = kotlin.random.Random.nextLong(),
  ): List<LibraryTrack> {
    val random = SplitMix64(seed)
    return tracks
      .map { track ->
        val randomValue = max(random.nextDouble(), Double.MIN_VALUE)
        val weight = selectionWeight(track, recentTrackIds)
        WeightedTrack(track, ln(randomValue) / weight)
      }
      .sortedWith(
        compareByDescending<WeightedTrack> { it.priority }
          .thenBy { it.track.id },
      )
      .take(maximumTrackCount.coerceAtLeast(0))
      .map { it.track }
  }

  fun selectionWeight(
    track: LibraryTrack,
    recentTrackIds: Set<String>,
  ): Double {
    val ageInYears = track.addedAt?.let { addedAt ->
      max(0.0, Duration.between(addedAt, referenceDate).seconds / SECONDS_PER_YEAR)
    } ?: 0.0
    val ageWeight = 1 + min(ageInYears, 12.0)
    val rotationWeight = if (track.id in recentTrackIds) 0.2 else 1.0
    return ageWeight * rotationWeight
  }

  private data class WeightedTrack(
    val track: LibraryTrack,
    val priority: Double,
  )

  private class SplitMix64(seed: Long) {
    private var state = seed.toULong()

    fun nextDouble(): Double {
      val value = next() shr 11
      return value.toDouble() / (1UL shl 53).toDouble()
    }

    private fun next(): ULong {
      state += 0x9E3779B97F4A7C15UL
      var value = state
      value = (value xor (value shr 30)) * 0xBF58476D1CE4E5B9UL
      value = (value xor (value shr 27)) * 0x94D049BB133111EBUL
      return value xor (value shr 31)
    }
  }

  private companion object {
    const val SECONDS_PER_YEAR = 365.25 * 24 * 60 * 60
  }
}
